### ISSUE TITLE
FileDialog on linux can't handle path elements for which the user has -r+x permissions

### ISSUE THREAD

**Issue by rowanworth** [fyne-io/fyne#4110]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

The environment I'm testing in has directories setup like along these lines:

```
HOME=/global/home/rowanworth

PERMISSIONS   OWNER        GROUP   PATH
 drwxr-xr-x   root         staff   /global
 drwx--x---   root         staff   /global/home
 drwx--x---   rowanworth   staff   /global/home/rowanworth
```

ie. my user `rowanworth`, who is a member of group `staff` has full read/write/execute permissions to my home directory `/global/home/rowanworth`. However I do not have read permissions on the parent `/global/home` directory, only execute permissions. This means the kernel VFS layer allows me to traverse to sub-directories of `/global/home/`, but not list its contents.

Within Fyne, this manifests as an inability to navigate the file dialog to files in my home directory. The default view upon opening the file picker shows a blank area where the grid of icons should be, and the path is displayed as:

```
/   global   home
```

If I toggle between the list/grid view then the files in my home directory _do_ appear - I can also navigate into subdirectories, but I have to toggle the list/grid view every time I change directories, and the path display never goes deeper than `/   global   home`.

### How to reproduce

1. Setup a directory that you have execute but not read permissions for
2. Try to navigate it in Fyne's file picker

### Example code

The environment is the critical component here rather than the code so I haven't tested this snippet but it should be close to working :)

```
import (
	"fmt"
	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/dialog"
	"fyne.io/fyne/v2/theme"
	"fyne.io/fyne/v2/widget"
)

func main() {
	a := app.New()
	win := a.NewWindow("Hello file picker")

	fileBrowser := dialog.NewFileOpen(func(u fyne.URIReadCloser, e error) { fmt.Println(u.String()); fmt.Println(e) }, win)
	win.SetContent(widget.NewButtonWithIcon("", theme.FolderOpenIcon(), func() { fileBrowser.Show() } ))

	win.ShowAndRun()
}
```

### Fyne version

v2@v2.3.5

### Go compiler version

go version go1.18.1 linux/amd64

### Operating system and version

Linux Centos 7.7.1908

### Additional Information

I'm very new to Fyne so no clue if this should be filed against the FileDialog or some deeper VFS interface.

**Comment by andydotxyz** [comment]
Were there no errors listed in your console?

**Comment by andydotxyz** [comment]
> The default view upon opening the file picker shows a blank area where the grid of icons should be, and the path is displayed as:

```
/   global   home
```

It sounds like maybe it's not getting your home directory correct perhaps - as it should be in the path... maybe a screenshot would help?

**Comment by rowanworth** [comment]
This is the only thing I see in the console:

```
2023/08/02 11:02:54 Fyne error:  Getting favorite locations
2023/08/02 11:02:54   Cause: this computer does not define a VIDEOS folder
2023/08/02 11:02:54   At: /localData/rowanw/go/pkg/mod/fyne.io/fyne/v2@v2.3.5/dialog/file.go:303
```

However I believe this relates to the "favorite folders" list and is unrelated to painting the current directory/file list (`xdg-user-dir VIDEO` returns the same path as my home folder, which causes `getFavoriteLocation()` in `dialog/file_xdg.go` to return this non-fatal error).

Here's a simple recipe to replicate the scenario in /tmp:

```
$ mkdir -p /tmp/global/home/rowanworth
$ chmod -r /tmp/global/home
$ export HOME=/tmp/global/home/rowanworth
$ echo hello world >$HOME/README.txt
$ mkdir $HOME/{Downloads,Documents,Music,Pictures,Videos}
$ ./main &
```

Here's a gif of what browsing that looks like. Note that:

1. the "current directory" never goes deeper than `/  tmp  global`, even though the file picker _starts_ in `/tmp/global/rowanworth` (ie. $HOME)
2. no files are ever rendered until I switch between the list/grid view

![Peek 2023-08-02 11-03](https://github.com/fyne-io/fyne/assets/2315064/3cea9165-1010-4850-8937-e3099313da1f)

**Comment by DaanSelen** [comment]
I also have this:

```
2025/11/21 14:37:34 Fyne error:  Getting favorite locations
2025/11/21 14:37:34   Cause: this computer does not define a VIDEOS folder
2025/11/21 14:37:34   At: /home/dselen/.go/pkg/mod/fyne.io/fyne/v2@v2.7.1/dialog/file.go:396
```

**Comment by andydotxyz** [comment]
Do you have a Videos folder in the standard home location?

**Comment by andydotxyz** [comment]
> Here's a simple recipe to replicate the scenario in /tmp:
> 
> ```
> $ mkdir -p /tmp/global/home/rowanworth
> $ chmod -r /tmp/global/home
> $ export HOME=/tmp/global/home/rowanworth
> $ echo hello world >$HOME/README.txt
> $ mkdir $HOME/{Downloads,Documents,Music,Pictures,Videos}
> $ ./main &
> ```

Thanks for these steps. It has been improved since the issue was opened - the path now displays in the bar and parent is there (though it cannot go up). The error I see now is:

```
2025/11/21 22:00:22 Fyne error:  Failed to set directory
2025/11/21 22:00:22   Cause: open /tmp/global/home/: permission denied
2025/11/21 22:00:22   At: /home/andy/Code/Fyne/fyne/dialog/file.go:510
```

I think that is consistent with expectations now?

However the README.txt is not displayed so there remains a (perhaps more subtle?) bug here to resolve

**Comment by rowanworth** [comment]
> 2025/11/21 22:00:22   Cause: open /tmp/global/home/: permission denied

This makes sense; `ls` gets the same error in console when trying to list the parent dir. However existing directory elements can be listed if their names are given:

```
$ cd $HOME
$ cd ..
$ ls
ls: cannot open directory '.': Permission denied

$ ls -ld rowanworth
drwx--x--- 2 rowanworth staff 4096 Nov 24 09:24 rowanworth

$ ls -ld .
drwx--x--- 2 root staff 4096 Nov 20 12:09 .
```

ie. access to my $HOME is unrestricted, but I can't trivially list the parent dir to enumerate the home directories of other users.

**Comment by andydotxyz** [comment]
You're quite right - I was in the wrong directory.

This is what I see now - it seems to be resolved :) please check out the `develop` branch and see if it is working for you too.

<img width="648" height="482" alt="Image" src="https://github.com/user-attachments/assets/3c439a02-55a2-4b06-bf05-531ac7ab4cd3" />

**Comment by rowanworth** [comment]
I haven't touched fyne in a long time; testing this is a long way off my critical path. I'll try to make time to check it out but don't be too surprised if it doesn't happen :)

Your screenshot looks like a great improvement, thanks for coming back to follow this one up!

**Comment by andydotxyz** [comment]
Not a problem. I'll assume this is fixed then as it seems to work for me.
Feel free to re-open if your testing shows otherwise.

**Comment by DaanSelen** [comment]
> Do you have a Videos folder in the standard home location?

No I do not. Its a headless system (WSL)

**Comment by andydotxyz** [comment]
Then it seems like the error is correct?

**Comment by DaanSelen** [comment]
> Then it seems like the error is correct?

True, but on a headless system. How to disable this? Because I don't want it to look for favorites.

**Comment by andydotxyz** [comment]
Please open a new issue and describe the request separately to this bug report.
